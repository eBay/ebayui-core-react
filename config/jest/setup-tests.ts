import '@testing-library/jest-dom'

declare var global: typeof globalThis;

if (typeof window.URL.createObjectURL === 'undefined') {
  Object.defineProperty(window.URL, 'createObjectURL', { value: () => {} });
}
