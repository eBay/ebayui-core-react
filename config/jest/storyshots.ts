import { render } from "@testing-library/react";
import initStoryshotsAddon from "@storybook/addon-storyshots";
import type { StoryshotsOptions } from "@storybook/addon-storyshots/dist/ts3.9/api/StoryshotsOptions";

const reactTestingLibrarySerializer = {
    print: (val, serialize, indent) => serialize(val.asFragment()),
    test: (val) => val && val.hasOwnProperty("container"),
};

export function initStoryshots(config: StoryshotsOptions) {
    initStoryshotsAddon({
        ...config,
        renderer: render,
        snapshotSerializers: [reactTestingLibrarySerializer],
    });
}
