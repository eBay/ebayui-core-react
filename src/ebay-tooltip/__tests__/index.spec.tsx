import React from 'react'
import requireContext from 'node-require-context'
import { render, fireEvent } from '@testing-library/react'
import { EbayTooltip, EbayTooltipContent, EbayTooltipHost } from '../index'
import { initStoryshots } from '../../../config/jest/storyshots'

jest.useFakeTimers()

const renderComponent = (props = {}) =>
    render(
        <EbayTooltip {...props}>
            <EbayTooltipHost>
                <button>Info</button>
            </EbayTooltipHost>
            <EbayTooltipContent>
                <p>Info content</p>
            </EbayTooltipContent>
        </EbayTooltip>
    )

const checkIsExpanded = (wrapper) => {
    expect(wrapper.container.querySelector('.tooltip--expanded')).toBeInTheDocument()
    expect(wrapper.container.querySelector('button[aria-expanded=true]')).toBeInTheDocument()
}

const checkIsCollapsed = (wrapper) => {
    expect(wrapper.container.querySelector('.tooltip--expanded')).not.toBeInTheDocument()
    expect(wrapper.container.querySelector('button[aria-expanded=true]')).not.toBeInTheDocument()
}

describe('<EbayTooltip>', () => {
    describe('on tooltip mouse enter', () => {
        it('should fire an event', () => {
            const spy = jest.fn()
            const { getByText } = renderComponent({ onMouseEnter: spy })
            fireEvent.mouseEnter(getByText('Info content'))

            expect(spy).toBeCalled()
        })

        it('should expand the tooltip', () => {
            const wrapper = renderComponent()
            fireEvent.mouseEnter(wrapper.getByText('Info content'))
            checkIsExpanded(wrapper)
        })

        it('should not expand the tooltip with noHover', () => {
            const wrapper = renderComponent({ noHover: true })
            fireEvent.mouseEnter(wrapper.getByText('Info content'))
            checkIsCollapsed(wrapper)
        })
    })

    describe('on tooltip mouse leave', () => {
        it('should fire an event', () => {
            const spy = jest.fn()
            const wrapper = renderComponent({ onMouseLeave: spy })
            fireEvent.mouseLeave(wrapper.getByText('Info content'))

            expect(spy).toBeCalled()
        })

        it('should collapse the tooltip', () => {
            const wrapper = renderComponent()
            fireEvent.mouseLeave(wrapper.getByText('Info content'))
            jest.runAllTimers()
            checkIsCollapsed(wrapper)
        })

        it('should not collapse the tooltip with noHover', () => {
            const wrapper = renderComponent({ noHover: true })
            const content = wrapper.getByText('Info content')
            fireEvent.focus(content)
            fireEvent.mouseLeave(content)
            checkIsExpanded(wrapper)
        })
    })

    describe('on tooltip focus', () => {
        it('should fire an event', () => {
            const spy = jest.fn()
            const wrapper = renderComponent({ onFocus: spy })
            fireEvent.focus(wrapper.getByText('Info'))

            expect(spy).toBeCalled()
        })

        it('should expand the tooltip', () => {
            const wrapper = renderComponent()
            fireEvent.focus(wrapper.getByText('Info'))
            checkIsExpanded(wrapper)
        })

        it('should expand the tooltip even with noHover', () => {
            const wrapper = renderComponent({ noHover: true })
            fireEvent.focus(wrapper.getByText('Info'))
            checkIsExpanded(wrapper)
        })
    })

    describe('on tooltip blur', () => {
        it('should fire an event', () => {
            const spy = jest.fn()
            const wrapper = renderComponent({ onBlur: spy })
            fireEvent.blur(wrapper.getByText('Info'))

            expect(spy).toBeCalled()
        })

        it('should collapse the tooltip', () => {
            const wrapper = renderComponent()
            fireEvent.blur(wrapper.getByText('Info'))
            checkIsCollapsed(wrapper)
        })
    })

    describe('on tooltip expanded', () => {
        it('should fire an event', () => {
            const spy = jest.fn()
            const wrapper = renderComponent({ onExpand: spy })
            fireEvent.focus(wrapper.getByText('Info'))

            expect(spy).toBeCalled()
        })
    })

    describe('on tooltip collapsed', () => {
        it('should fire an event', () => {
            const spy = jest.fn()
            const wrapper = renderComponent({ onCollapse: spy })
            fireEvent.blur(wrapper.getByText('Info'))

            expect(spy).toBeCalled()
        })
    })

    describe('on using the tooltip with no host', () => {
        it('should throw an error', () => {
            jest.spyOn(console, 'error').mockImplementation(() => null)
            expect(() => {
                render(
                    <EbayTooltip>
                        <button>Test</button>
                    </EbayTooltip>
                )
            }).toThrowError(`EbayTooltip: Please use a EbayTooltipHost that defines the host of the tooltip`)
            ;(console.error as jest.Mock).mockRestore()
        })
    })

    describe('on using the tooltip with no content', () => {
        it('should throw an error', () => {
            jest.spyOn(console, 'error').mockImplementation(() => null)
            expect(() => {
                render(
                    <EbayTooltip>
                        <EbayTooltipHost>
                            <a>Test</a>
                        </EbayTooltipHost>
                    </EbayTooltip>
                )
            }).toThrowError(`EbayTooltip: Please use a EbayTooltipContent that defines the content of the tooltip`)
            ;(console.error as jest.Mock).mockRestore()
        })
    })

    describe('on using more than one element in the host', () => {
        it('should throw an error', () => {
            jest.spyOn(console, 'error').mockImplementation(() => null)
            expect(() => {
                render(
                    <EbayTooltip>
                        <EbayTooltipHost>
                            <a>Test</a>
                            <button>Test</button>
                        </EbayTooltipHost>
                        <EbayTooltipContent>Content</EbayTooltipContent>
                    </EbayTooltip>
                )
            }).toThrowError(`React.Children.only expected to receive a single React element child.`)
            ;(console.error as jest.Mock).mockRestore()
        })
    })
})

initStoryshots({
    config: ({ configure }) => {
        const req = requireContext('./', false, /\.stories\.tsx$/);
        return configure(req, module)
    }

});
