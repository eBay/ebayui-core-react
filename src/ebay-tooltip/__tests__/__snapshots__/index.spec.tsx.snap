// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots ebay-tooltip Anchor host 1`] = `
<DocumentFragment>
  <div
    style="width: 100%; text-align: center; margin-top: 100px;"
  >
    <span
      class="tooltip"
    >
      <a
        aria-expanded="false"
        class="tooltip__host"
        href="https://www.ebay.com"
      >
        View options
      </a>
      <span
        class="tooltip__overlay"
        role="tooltip"
        style="transform: translateX(-50%); left: 50%;"
      >
        <span
          class="tooltip__pointer tooltip__pointer--bottom"
        />
        <span
          class="tooltip__mask"
        >
          <span
            class="tooltip__cell"
          >
            <span
              class="tooltip__content"
            >
              <p>
                Use Access Key 'S' to display settings.
              </p>
            </span>
          </span>
        </span>
      </span>
    </span>
  </div>
</DocumentFragment>
`;

exports[`Storyshots ebay-tooltip Default tooltip 1`] = `
<DocumentFragment>
  <div
    style="width: 100%; text-align: center; margin-top: 100px;"
  >
    <span
      class="tooltip"
    >
      <button
        aria-expanded="false"
        class="btn tooltip__host btn--secondary"
      >
        Info
      </button>
      <span
        class="tooltip__overlay"
        role="tooltip"
        style="transform: translateX(-50%); left: 50%;"
      >
        <span
          class="tooltip__pointer tooltip__pointer--bottom"
        />
        <span
          class="tooltip__mask"
        >
          <span
            class="tooltip__cell"
          >
            <span
              class="tooltip__content"
            >
              <p>
                Content
              </p>
            </span>
          </span>
        </span>
      </span>
    </span>
  </div>
</DocumentFragment>
`;

exports[`Storyshots ebay-tooltip No hover 1`] = `
<DocumentFragment>
  <div
    style="width: 100%; text-align: center; margin-top: 100px;"
  >
    <span
      class="tooltip"
    >
      <span
        class="textbox"
      >
        <input
          aria-expanded="false"
          class="textbox__control"
          placeholder="Email address"
          type="text"
          value=""
        />
      </span>
      <span
        class="tooltip__overlay"
        role="tooltip"
        style="transform: translateX(-50%); left: 50%;"
      >
        <span
          class="tooltip__pointer tooltip__pointer--bottom"
        />
        <span
          class="tooltip__mask"
        >
          <span
            class="tooltip__cell"
          >
            <span
              class="tooltip__content"
            >
              <p>
                Use Access Key 'S' to display settings.
              </p>
            </span>
          </span>
        </span>
      </span>
    </span>
  </div>
</DocumentFragment>
`;

exports[`Storyshots ebay-tooltip Pointer direction 1`] = `
<DocumentFragment>
  <div>
    <div
      style="margin: 100px 0px; width: 100%; text-align: center;"
    >
      <span
        class="tooltip"
      >
        <a
          aria-expanded="false"
          class="tooltip__host"
          href="https://www.ebay.com"
        >
          top
        </a>
        <span
          class="tooltip__overlay"
          role="tooltip"
          style="transform: translateX(-50%); left: 50%;"
        >
          <span
            class="tooltip__pointer tooltip__pointer--top"
          />
          <span
            class="tooltip__mask"
          >
            <span
              class="tooltip__cell"
            >
              <span
                class="tooltip__content"
              >
                <p>
                  Use Access Key 'S' to display settings.
                </p>
              </span>
            </span>
          </span>
        </span>
      </span>
    </div>
    <div
      style="margin: 100px 0px; width: 100%; text-align: center;"
    >
      <span
        class="tooltip"
      >
        <a
          aria-expanded="false"
          class="tooltip__host"
          href="https://www.ebay.com"
        >
          top-left
        </a>
        <span
          class="tooltip__overlay"
          role="tooltip"
        >
          <span
            class="tooltip__pointer tooltip__pointer--top-left"
          />
          <span
            class="tooltip__mask"
          >
            <span
              class="tooltip__cell"
            >
              <span
                class="tooltip__content"
              >
                <p>
                  Use Access Key 'S' to display settings.
                </p>
              </span>
            </span>
          </span>
        </span>
      </span>
    </div>
    <div
      style="margin: 100px 0px; width: 100%; text-align: center;"
    >
      <span
        class="tooltip"
      >
        <a
          aria-expanded="false"
          class="tooltip__host"
          href="https://www.ebay.com"
        >
          top-right
        </a>
        <span
          class="tooltip__overlay"
          role="tooltip"
          style="right: -7px;"
        >
          <span
            class="tooltip__pointer tooltip__pointer--top-right"
          />
          <span
            class="tooltip__mask"
          >
            <span
              class="tooltip__cell"
            >
              <span
                class="tooltip__content"
              >
                <p>
                  Use Access Key 'S' to display settings.
                </p>
              </span>
            </span>
          </span>
        </span>
      </span>
    </div>
    <div
      style="margin: 100px 0px; width: 100%; text-align: center;"
    >
      <span
        class="tooltip"
      >
        <a
          aria-expanded="false"
          class="tooltip__host"
          href="https://www.ebay.com"
        >
          right
        </a>
        <span
          class="tooltip__overlay"
          role="tooltip"
          style="transform: translateX(-16px) translateY(-50%); right: 100%; top: -4px;"
        >
          <span
            class="tooltip__pointer tooltip__pointer--right"
          />
          <span
            class="tooltip__mask"
          >
            <span
              class="tooltip__cell"
            >
              <span
                class="tooltip__content"
              >
                <p>
                  Use Access Key 'S' to display settings.
                </p>
              </span>
            </span>
          </span>
        </span>
      </span>
    </div>
    <div
      style="margin: 100px 0px; width: 100%; text-align: center;"
    >
      <span
        class="tooltip"
      >
        <a
          aria-expanded="false"
          class="tooltip__host"
          href="https://www.ebay.com"
        >
          right-bottom
        </a>
        <span
          class="tooltip__overlay"
          role="tooltip"
          style="transform: translateX(-16px); right: 100%; bottom: -50%;"
        >
          <span
            class="tooltip__pointer tooltip__pointer--right-bottom"
          />
          <span
            class="tooltip__mask"
          >
            <span
              class="tooltip__cell"
            >
              <span
                class="tooltip__content"
              >
                <p>
                  Use Access Key 'S' to display settings.
                </p>
              </span>
            </span>
          </span>
        </span>
      </span>
    </div>
    <div
      style="margin: 100px 0px; width: 100%; text-align: center;"
    >
      <span
        class="tooltip"
      >
        <a
          aria-expanded="false"
          class="tooltip__host"
          href="https://www.ebay.com"
        >
          right-top
        </a>
        <span
          class="tooltip__overlay"
          role="tooltip"
          style="transform: translateX(-16px); right: 100%; top: -100%;"
        >
          <span
            class="tooltip__pointer tooltip__pointer--right-top"
          />
          <span
            class="tooltip__mask"
          >
            <span
              class="tooltip__cell"
            >
              <span
                class="tooltip__content"
              >
                <p>
                  Use Access Key 'S' to display settings.
                </p>
              </span>
            </span>
          </span>
        </span>
      </span>
    </div>
    <div
      style="margin: 100px 0px; width: 100%; text-align: center;"
    >
      <span
        class="tooltip"
      >
        <a
          aria-expanded="false"
          class="tooltip__host"
          href="https://www.ebay.com"
        >
          bottom
        </a>
        <span
          class="tooltip__overlay"
          role="tooltip"
          style="transform: translateX(-50%); left: 50%;"
        >
          <span
            class="tooltip__pointer tooltip__pointer--bottom"
          />
          <span
            class="tooltip__mask"
          >
            <span
              class="tooltip__cell"
            >
              <span
                class="tooltip__content"
              >
                <p>
                  Use Access Key 'S' to display settings.
                </p>
              </span>
            </span>
          </span>
        </span>
      </span>
    </div>
    <div
      style="margin: 100px 0px; width: 100%; text-align: center;"
    >
      <span
        class="tooltip"
      >
        <a
          aria-expanded="false"
          class="tooltip__host"
          href="https://www.ebay.com"
        >
          bottom-left
        </a>
        <span
          class="tooltip__overlay"
          role="tooltip"
          style="left: -7px;"
        >
          <span
            class="tooltip__pointer tooltip__pointer--bottom-left"
          />
          <span
            class="tooltip__mask"
          >
            <span
              class="tooltip__cell"
            >
              <span
                class="tooltip__content"
              >
                <p>
                  Use Access Key 'S' to display settings.
                </p>
              </span>
            </span>
          </span>
        </span>
      </span>
    </div>
    <div
      style="margin: 100px 0px; width: 100%; text-align: center;"
    >
      <span
        class="tooltip"
      >
        <a
          aria-expanded="false"
          class="tooltip__host"
          href="https://www.ebay.com"
        >
          bottom-right
        </a>
        <span
          class="tooltip__overlay"
          role="tooltip"
          style="right: -7px;"
        >
          <span
            class="tooltip__pointer tooltip__pointer--bottom-right"
          />
          <span
            class="tooltip__mask"
          >
            <span
              class="tooltip__cell"
            >
              <span
                class="tooltip__content"
              >
                <p>
                  Use Access Key 'S' to display settings.
                </p>
              </span>
            </span>
          </span>
        </span>
      </span>
    </div>
    <div
      style="margin: 100px 0px; width: 100%; text-align: center;"
    >
      <span
        class="tooltip"
      >
        <a
          aria-expanded="false"
          class="tooltip__host"
          href="https://www.ebay.com"
        >
          left
        </a>
        <span
          class="tooltip__overlay"
          role="tooltip"
          style="transform: translateX(16px) translateY(-50%); left: 100%; top: -5px;"
        >
          <span
            class="tooltip__pointer tooltip__pointer--left"
          />
          <span
            class="tooltip__mask"
          >
            <span
              class="tooltip__cell"
            >
              <span
                class="tooltip__content"
              >
                <p>
                  Use Access Key 'S' to display settings.
                </p>
              </span>
            </span>
          </span>
        </span>
      </span>
    </div>
    <div
      style="margin: 100px 0px; width: 100%; text-align: center;"
    >
      <span
        class="tooltip"
      >
        <a
          aria-expanded="false"
          class="tooltip__host"
          href="https://www.ebay.com"
        >
          left-bottom
        </a>
        <span
          class="tooltip__overlay"
          role="tooltip"
          style="transform: translateX(16px); left: 100%; bottom: -7px;"
        >
          <span
            class="tooltip__pointer tooltip__pointer--left-bottom"
          />
          <span
            class="tooltip__mask"
          >
            <span
              class="tooltip__cell"
            >
              <span
                class="tooltip__content"
              >
                <p>
                  Use Access Key 'S' to display settings.
                </p>
              </span>
            </span>
          </span>
        </span>
      </span>
    </div>
    <div
      style="margin: 100px 0px; width: 100%; text-align: center;"
    >
      <span
        class="tooltip"
      >
        <a
          aria-expanded="false"
          class="tooltip__host"
          href="https://www.ebay.com"
        >
          left-top
        </a>
        <span
          class="tooltip__overlay"
          role="tooltip"
          style="transform: translateX(16px); left: 100%; top: -100%;"
        >
          <span
            class="tooltip__pointer tooltip__pointer--left-top"
          />
          <span
            class="tooltip__mask"
          >
            <span
              class="tooltip__cell"
            >
              <span
                class="tooltip__content"
              >
                <p>
                  Use Access Key 'S' to display settings.
                </p>
              </span>
            </span>
          </span>
        </span>
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Storyshots ebay-tooltip Pointer with custom location 1`] = `
<DocumentFragment>
  <div
    style="width: 100%; text-align: center; margin-top: 100px;"
  >
    <span
      class="tooltip"
    >
      <a
        aria-expanded="false"
        class="tooltip__host"
        href="https://www.ebay.com"
      >
        View options
      </a>
      <span
        class="tooltip__overlay"
        role="tooltip"
        style="top: 40px; left: -16px;"
      >
        <span
          class="tooltip__pointer tooltip__pointer--top-left"
        />
        <span
          class="tooltip__mask"
        >
          <span
            class="tooltip__cell"
          >
            <span
              class="tooltip__content"
            >
              <p>
                Use Access Key 'S' to display settings.
              </p>
            </span>
          </span>
        </span>
      </span>
    </span>
  </div>
</DocumentFragment>
`;
